pkg/flow/internal/controller/node_config_import.go

```diff
@@ -0,0 +1,328 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"go.uber.org/atomic"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/pkg/flow/internal/importsource"
+	"github.com/grafana/agent/pkg/flow/logging/level"
+	"github.com/grafana/agent/pkg/flow/tracing"
+	"github.com/grafana/river/ast"
+	"github.com/grafana/river/parser"
+	"github.com/grafana/river/vm"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// ImportConfigNode imports declare and import blocks via a managed import source.
+// The imported declare are stored in importedDeclares.
+// For every imported import block, the ImportConfigNode will create ImportConfigNode children.
+// The children are evaluated and ran by the parent.
+// When an ImportConfigNode receives new content from its source, it updates its importedDeclares and recreates its children.
+// Then an update call is propagated to the root ImportConfigNode to inform the controller for reevaluation.
+type ImportConfigNode struct {
+	nodeID        string
+	globalID      string
+	label         string
+	componentName string
+	globals       ComponentGlobals          // Need a copy of the globals to create other import nodes
+	block         *ast.BlockStmt            // Current River blocks to derive config from
+	source        importsource.ImportSource // source retrieves the module content
+
+	OnBlockNodeUpdate func(cn BlockNode) // notifies the controller or the parent for reevaluation
+	logger            log.Logger
+
+	importChildrenUpdateChan chan struct{} // used to trigger an update of the running children
+
+	mut                       sync.RWMutex
+	importedContent           string
+	importConfigNodesChildren map[string]*ImportConfigNode
+	importChildrenRunning     bool
+	importedDeclares          map[string]ast.Body
+
+	inContentUpdate atomic.Bool
+}
+
+var _ RunnableNode = (*ImportConfigNode)(nil)
+
+// NewImportConfigNode creates a new ImportConfigNode from an initial ast.BlockStmt.
+// The underlying config isn't applied until Evaluate is called.
+func NewImportConfigNode(block *ast.BlockStmt, globals ComponentGlobals, sourceType importsource.SourceType) *ImportConfigNode {
+	nodeID := BlockComponentID(block).String()
+
+	globalID := nodeID
+	if globals.ControllerID != "" {
+		globalID = path.Join(globals.ControllerID, nodeID)
+	}
+
+	cn := &ImportConfigNode{
+		nodeID:                   nodeID,
+		globalID:                 globalID,
+		label:                    block.Label,
+		componentName:            block.GetBlockName(),
+		globals:                  globals,
+		block:                    block,
+		OnBlockNodeUpdate:        globals.OnBlockNodeUpdate,
+		importChildrenUpdateChan: make(chan struct{}),
```

Comment:
Should this be a buffered channel so writes don't block? `make(chan struct{}, 1)`